### ISSUE TITLE
While resetting DataLoader, RuntimeError: Pin memory thread exited unexpectedly

### ISSUE THREAD

**Issue by gibiansky** [pytorch/pytorch#47445]
## 🐛 Bug

The following exception is raised:

```
RuntimeError: Pin memory thread exited unexpectedly
```

This is happening at the end of an epoch (using PyTorch Lightning). Here's a (poorly formatted) traceback:

```
    for batch_idx, (batch, is_last_batch) in train_dataloader:
  File ".../pytorch_lightning/profiler/profilers.py", line 80, in
 profile_iterable
    value = next(iterator)
  File ".../pytorch_lightning/trainer/connectors/data_connector.py
", line 45, in _with_is_last
    it = iter(iterable)
  File ".../torch/utils/data/dataloader.py", line 356, in __iter__
    self._iterator._reset(self)
  File ".../torch/utils/data/dataloader.py", line 936, in _reset
    data = self._get_data()
  File ".../torch/utils/data/dataloader.py", line 1113, in _get_data
    raise RuntimeError('Pin memory thread exited unexpectedly')
RuntimeError: Pin memory thread exited unexpectedly
```

Since it's at the end of an epoch, the data loader is being reset (see `self._iterator._reset()` in the traceback), and for reasons I can't understand, it causes issues with the pin memory thread.

This of course happens with pin_memory=True (as well as `num_workers=8, prefetch_factor=8, persistent_workers=True`). I have a suspicion it might be due to persistent_workers=True.

## Environment

I unfortunately don't have access to easily run the environment collection script. I will figure this out if it's critical; however, it's a very recently built version of PyTorch running on Linux with V100 GPUs.

cc @SsnL @VitalyFedyunin @ejguan

**Comment by glaringlee** [comment]
@gibiansky
Is that possible for you to provide some sample code to reproduce it? Thanks.

**Comment by Mmdixon** [comment]
@glaringlee I was able to reproduce it with this:
```python
#!/usr/bin/env python

import torch
from torch.utils.data import DataLoader, IterableDataset

data_size = 100
batch_size = 5
worker_size = 5


class MyIterDataset(IterableDataset):
    def __iter__(self):
        for i in range(data_size):
            yield torch.ones([10])


def test_bug():
    loader = DataLoader(MyIterDataset(),
                        batch_size=batch_size,
                        drop_last=True,
                        num_workers=worker_size,
                        pin_memory=True,
                        persistent_workers=True)
    for epoch in range(2):
        print("EPOCH", epoch)
        for batch in loader:
            pass
    print("DONE")


if __name__ == "__main__":
    test_bug()
```

```bash
# ./bug_repo.py 
EPOCH 0
EPOCH 1
Exception in thread Thread-1:
Traceback (most recent call last):
  File "/miniconda/lib/python3.7/threading.py", line 926, in _bootstrap_inner
    self.run()
  File "/miniconda/lib/python3.7/threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
  File "/miniconda/lib/python3.7/site-packages/torch/utils/data/_utils/pin_memory.py", line 28, in _pin_memory_loop
    idx, data = r
ValueError: not enough values to unpack (expected 2, got 0)

Traceback (most recent call last):
  File "./bug_repo.py", line 32, in <module>
    test_bug()
  File "./bug_repo.py", line 26, in test_bug
    for batch in loader:
  File "/miniconda/lib/python3.7/site-packages/torch/utils/data/dataloader.py", line 349, in __iter__
    self._iterator._reset(self)
  File "/miniconda/lib/python3.7/site-packages/torch/utils/data/dataloader.py", line 852, in _reset
    data = self._get_data()
  File "/miniconda/lib/python3.7/site-packages/torch/utils/data/dataloader.py", line 1029, in _get_data
    raise RuntimeError('Pin memory thread exited unexpectedly')
RuntimeError: Pin memory thread exited unexpectedly

# Setting pin_memory=False
# ./bug_repo.py 
EPOCH 0
EPOCH 1
DONE
```

**Comment by jpetkau** [comment]
It looks like `persistent_workers` just doesn't work with `pin_memory`. Reproduced this on Windows with pytorch 1.7.0, and on Linux with 1.8.0, with both indexed and iterable datasets.

Here's a minimal repro:
```
import torch

if __name__ == '__main__':
    loader = torch.utils.data.DataLoader(
        dataset=range(10), num_workers=2,
        pin_memory=True, persistent_workers=True,
    )
    for x in loader: print(x)
    for x in loader: print(x)
```

```
Exception in thread Thread-1:
Traceback (most recent call last):
  File "...\Python38\lib\threading.py", line 932, in _bootstrap_inner
    self.run()
  File "...\Python38\lib\threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
  File "...\Python38\lib\site-packages\torch\utils\data\_utils\pin_memory.py", line 28, in _pin_memory_loop
    idx, data = r
ValueError: not enough values to unpack (expected 2, got 0)
Traceback (most recent call last):
  File "dataloader_test.py", line 9, in <module>
    for x in loader: ...
  File "...\Python38\lib\site-packages\torch\utils\data\dataloader.py", line 349, in __iter__
    self._iterator._reset(self)
  File "...\Python38\lib\site-packages\torch\utils\data\dataloader.py", line 852, in _reset
    data = self._get_data()
  File "...\Python38\lib\site-packages\torch\utils\data\dataloader.py", line 1029, in _get_data
    raise RuntimeError('Pin memory thread exited unexpectedly')
RuntimeError: Pin memory thread exited unexpectedly
```

**Comment by ejguan** [comment]
Closing as the issue as solved in #48543

**Comment by ejguan** [comment]
@GonyRosenman 
I am assuming you are using 1.7.1 version. You can use pytorch-nightly to get rid of this error as the patch has already landed into master.
